README.md
pixenate
========

## Build instructions

    perl makefile.pl package

... Creates a tar.gz file in the ./target directory . Unzip the .tar.gz into the directory where you want to install Pixenate.

    


releasenotes.txt
Pixenate Release Notes
======================

Release notes are displayed in reverse chronological order.

The Installation Guide can be found at...

http://pixenate.com/pixenate/docs/Installation-Guide.html

The Installation Guide (and all other documenation) is also included
in the Pixenate tar.gz file.

20110531
--------
Add support for https (secure http). (changed upload.pl and Fetch.pm modules)

20110205
--------
Changed issconfig.gif - Added "" (doublequotes) around perl parameters
(for case of folders with spaces in them on Windows)

20110202
--------
+----------------------------------------------------------------------+
| Version                                                              |
| 1.3.5sp3                                                             |
+----------------------------------------------------------------------+
Fix for security bug in save.pl & upload.pl 
(sanitize image & filename parameters before opening) 

20101212
--------
+----------------------------------------------------------------------+
| Version                                                              |
| 1.3.5sp2                                                             |
+----------------------------------------------------------------------+

Miscellaneous bug fixes.

20100119
--------
Added new example file example-minimum-resolution.html which
demonstrates limiting the selection area based on a product chosen
from a drop-down list (common in photo-merchandising).

20091221
--------
Bug Fix: canvas not initialized correctly when inside a FancyBox container.
Changed file: pxn8_dom.js

20091004
--------
Added new zoom helper functions PXN8.zoom.toWidth()
PNX8.zoom.shrinkToWidth() PXN8.zoom.expandToWidth()
PXN8.zoom.toHeight() PXN8.zoom.shrinkToHeight() PXN8.zoom.expandToHeight()

20091002
--------
+----------------------------------------------------------------------+
| Version                                                              |
| 1.3.5sp1                                                             |
+----------------------------------------------------------------------+

20090827
--------
Bug Fix: Freehand Drawing did not work correctly with the BEEP
plugin. This has been fixed.

20090826
--------
Added new example-text-placement page which demonstrates how to place
text on the image using a combination of PXN8.ImageMagick and
PXN8.overlay functions.

20090802
--------
Bug Fix: Freehand drawing did not account for window scrolling. The
window.pageXOffset and window.pageYOffset attributes were ignored when
performing freehand drawing. This has been fixed.


20090726
--------
Bug Fix: Pixenate fails to load edited images if the path where
pixenate is installed contains a . (dot) character. This has been
fixed.

20090501
--------
+----------------------------------------------------------------------+
| Version                                                              |
| 1.3.5                                                                |
+----------------------------------------------------------------------+

20090430
--------
Bug Fix: Interoperability problems with Prototype 1.6. When operations
were wrapped inside a PXN8.tools.startTransaction() and
PXN8.tools.endTransaction(), and the prototype library was being used,
an error would appear when trying to edit the photos...

  The '' plugin is not available in this version of the software.

This problem has now been fixed.

20090430
--------
Added new PXN8.tools.freehand() tool to the 'slick' theme.

20090429
--------
Improved color picker used in Slick theme.

20090417
--------
Added new freehand drawing example page. /pixenate/docs/example-freehand.html

20090416
--------
Added PXN8.dom.css() function for convenient setting of CSS properties
on an element.

20090414
--------
Added PXN8.freehand.* module for freehand drawing on top of photos.
See updated API-Reference for more details.


20090407
--------
Improved Fetch.pm plugin. Now works better with IIS / mod_perl (file
test twice - once without and once with PIXENATE_WORKING_DIRECTORY
prefix).

20090327
--------
Added pixenate/fonts/ directory which includes a small sample of free
fonts which can be used in Pixenate. (also includes type-pixenate.xml
and instructions for installing the fonts for use with
ImageMagick/Pixenate )

20090324
--------
Update Installation guide. Appendix A added - Using Pixenate in a
secure website (SSL) with Virtual Directories.

20090320
--------
Bug Fix: Memphis / Hires Themes constrain the canvas to a fixed width
and height. The constraints did not work in IE when any DOCTYPE was
declared.
This has been fixed.
(see bug:http://mt-olympus.com/emmett/bug_overflow_positionrelative.php )


20090318
--------
Bug Fix: Spirit-level mode (as used in Slick theme) selected off-image
areas when the image was zoomed and then changed. This behaviour has
been fixed.

20090318 
--------
Improved initialization handling in PXN8.select and PXN8.zoom family
of functions. Now possible to write the following code without need
for PXN8.listener.onceOnly() calls...
   
   PXN8.initialize("/pixenate/images/samples/garinish.jpg");
	PXN8.selectByRatio("1x1");
	PXN8.zoom.zoomByValue(2);


20090317
--------
Changed PXN8.selectByRatio() to work in the following scenario...

   PXN8.initialize("/pixenate/images/samples/garinish.jpg");
   PXN8.selectByRatio("1x1");

Previously this would not have worked because the image would not have
yet loaded by the time PXN8.selectByRatio() was called.

20090305
--------
+----------------------------------------------------------------------+
| Version                                                              |
| 1.3.4sp1                                                             |

webroot/upload.pl
#!/usr/bin/perl -I./lib
#   
# ^^^
# The #! (hash-bang) line says which command to use to interpret this script.
# If perl is installed in a different location to the one above then change
# this line.
#

use lib "[% print $lib_dir; %]";
#
# ^^^
# This line says which directory to load supporting modules from. If you are
# running this script in a standard CGI environment (i.e. NOT mod_perl or perliis.dll)
# then you can leave this line unchanged.
# 
# If you ARE running this script in a non-CGI environment (either mod_perl or the 
# perliis.dll ISAPI filter on IIS) then you should change the value between quotes ("lib")
# to be the full path to the pixenate/lib directory. e.g.
#
# /usr/local/www/apache22/data/pixenate/lib 
# 
# or
#
# c:/inetpub/wwwroot/pixenate/lib
#

# !!! DO NOT CHANGE ANY CODE BELOW THIS LINE !!!
#-----------------------------------------------------------------------
#
# (c) Copyright 2005-2010 SXOOP Technologies Ltd. All rights reserved.
#
# upload.pl
# 
# Upload files to Pixenate's cache directory for editing
#
# ----------------------------------------------------------------------------
use strict;


use CGI ':standard';
use CGI::Carp qw(carpout fatalsToBrowser);
use Digest::MD5;
use File::Spec;

eval 
{
    require Sxoop::PXN8::Debug;
    require Sxoop::PXN8::Bootstrap;
    Sxoop::PXN8::Bootstrap::configure();
};
if ($@)
{
    my $error = "Unable to load server-side pixenate perl modules: Change the [use lib] statement in upload.pl";
    handle_error ("$error : $@");
}
else 
{
    #
    # try to handle the request
    #
    eval 
    {
	handle_request();
    };
    if ($@)
    {
	#
	# catch the error
	#
	handle_error($@);
    }
}
#
# 
#
sub handle_request 
{
    # ----------------------------
    # set file upload size limit
    # ----------------------------
    
    Sxoop::PXN8::Debug::log ("upload.pl> [info] setting CGI::POST_MAX to $ENV{PIXENATE_CGI_POST_MAX}\n");
    $CGI::POST_MAX = $ENV{PIXENATE_CGI_POST_MAX};
    
    Sxoop::PXN8::Debug::log ("upload.pl> [info] creating new CGI object\n");
    
    my $query = 0;
    
    eval 
    {
	$query = new CGI();
    };
    if ($@)
    {
	Sxoop::PXN8::Debug::log ("upload.pl> [error] $@\n");
	
	print STDOUT "Content-type: text/plain\n\n";
	print STDOUT "An error occurred while trying to upload the file...\n";
	print STDOUT "$@";
	return;
    }
    my $next_page = $query->param("next_page");
    my $mode = $query->param("mode");
	 
    if ($mode ne "api" && $next_page !~ /.+/)
    {
	Sxoop::PXN8::Debug::log ("upload.pl> [warn] next_page parameter is empty");
	print STDOUT "Content-type: text/plain\n\n";
	print STDOUT "next_page parameter is not present.\n";
	return;
	
    }
    my @param_names = $query->param();
    foreach(@param_names)
    {
	Sxoop::PXN8::Debug::log ("upload.pl> [param] $_");
    }
    
    my %passthru =  map { 
	my $v = $query->param($_); 
	$_ =~ s/^_//; $_=>$v;
    } grep /^_/, @param_names;
    
    Sxoop::PXN8::Debug::log ("upload.pl> [info] created new CGI object\n");
    
    my $upload_filehandle = $query->upload('filename');
    my $image_param_name = $query->param("image_param_name");
    my $hires_image_param_name = $query->param("hires_image_param_name");
    my $pxn8_root = $query->param("pxn8_root");
    my $max_dimension = $query->param("max_dim");
    my $filename = $query->param('filename');
    
    my $cgi_error = $query->cgi_error();
    Sxoop::PXN8::Debug::log ("upload.pl> [info] cgi_error = $cgi_error\n");
    
    if ($cgi_error =~ /^413/)
    {
	my $kbLimit = $ENV{PIXENATE_CGI_POST_MAX}/1024;
	
	print STDOUT "Content-type: text/html\n\n";
	print STDOUT "<html><body>\n";
	print STDOUT "You cannot upload images greater than $kbLimit Kilobytes in size!";
	print STDOUT "<br/>Click the browser's back button to continue...";
	print STDOUT "</body></html>";
	return;
    }
    elsif ($cgi_error =~ /.+/)
    {
	print STDOUT "Content-type: text/html\n\n";
	print STDOUT "<html><body>\n";
	print STDOUT "<p>Sorry - An error occurred while uploading $filename.</p> ";
	print STDOUT "<p>$cgi_error</p>\n";
	print STDOUT "</body></html>";
	return;
    }
    
    Sxoop::PXN8::Debug::log ("upload.pl> End-user uploaded file $filename\n");
    
#
# bug fix 20060721 save.pl should use original filename
# when saving image to disk.
#
	 my $original_filename = $filename; 
#
# wph 20070117: Fix IE bug on save-to-disk of uploaded images
# (IE won't save if the uploaded image has spaces in it's name)
#
    my @ofnParts = split /[\/\\]/, $original_filename;
    $original_filename = $ofnParts[$#ofnParts];
    $original_filename =~ s/\s/_/g;
    
	 
    my ($file_ext) = $filename =~ /(\.[a-zA-Z]+)$/;
    $filename = sprintf("photo_%s.jpg", Digest::MD5::md5_hex($filename . $query->remote_host() ));
    
    my $relative_filepath = "NO_FILE_UPLOADED";
#
# flag to indicate if image had to be resized
#
    my $resize = 0;
    
    if ($upload_filehandle)
    {
	
	#---------------------------------------
	# save the uploaded image to the server
	#---------------------------------------
		  
	$relative_filepath = "$ENV{PIXENATE_CACHE_DIR}/$filename";
	
# wph 20110202 sanitize the filename (remove | chars)
	$relative_filepath =~ s/\|.*//;

	my $absolute_filepath = File::Spec->rel2abs($relative_filepath,$ENV{PIXENATE_WORKING_DIRECTORY});
	
	Sxoop::PXN8::Debug::log ("upload.pl> [info] opening $absolute_filepath for output\n");
	
	if (! open UPLOADFILE, ">$absolute_filepath") 
	{
	    Sxoop::PXN8::Debug::log ("upload.pl> [error] cant open $absolute_filepath because $!\n");
	    die "Could not open file $absolute_filepath: $!\n";
	}
	
	binmode UPLOADFILE;
	while ( <$upload_filehandle> )
	{
	    print UPLOADFILE;
	}
	close UPLOADFILE;
	
	Sxoop::PXN8::Debug::log ("upload.pl> [info] saved to $absolute_filepath\n");
	
	#
	# wph 20070813 : 'use Image::Magick' caused CGI.pm to fail with the following error...
	#
	# CGI.pm: Server closed socket during multipart read (client aborted?).
	#
	# I assume this was because upload.pl is a mutlipart/form-data enctype and so 
	# CGI.pm must read from STDIN. 
	# Unfortunately there seems to be a conflict with CGI.pm version 3.2.9 and Image::Magick 6.3.3
	# I suspect that a 'use Image::Magick' will do something to the STDIN stream which causes CGI.pm
	# to fail.
	# The workaround is to 'require Image::Magick' - this means the imagemagick module is loaded at
	# runtime - not compile-time as with 'use'.
	#
	require Image::Magick;
	my $im = new Image::Magick();
	
	my $imres = $im->Read($absolute_filepath);
	if ($imres)
	{
	    my $error_msg = "Could not read file $absolute_filepath because $imres";
	    Sxoop::PXN8::Debug::log ("upload.pl> ERROR> $error_msg\n");
	    die $error_msg;
	}
	my $original_image = $im->Clone();
	
	if ($imres)
	{
	    # its not a valid image !!!
	    Sxoop::PXN8::Debug::log ("upload.pl> about to remove invalid uploaded image [$absolute_filepath]\n");
	    unlink $absolute_filepath or die "Could not remove invalid file $absolute_filepath: $!";
	    Sxoop::PXN8::Debug::log ("upload.pl> removed image\n");
	    die "Invalid file upload : File is not an image : $!\n";
	    
	}  
	if ($max_dimension)
	{
	    #
	    # must resize images that are too big to more manageable web dimensions
	    #
	    my $iw = $im->Get('width');
	    my $ih = $im->Get('height');
	    if ($iw > $ih)
	    {
		if ($iw > $max_dimension)
		{
		    $ih = $ih * ($max_dimension / $iw);
		    $iw = $max_dimension;
		    $resize = 1;
		}
	    }
	    else
	    {
		if ($ih > $max_dimension)
		{
		    $iw = $iw * ($max_dimension/ $ih);
		    $ih = $max_dimension;
		    $resize = 1;
		}
	    }
	    
	    if ($resize == 1)
	    {
		Sxoop::PXN8::Debug::log ("upload.pl> [info] resizing image to $iw x $ih\n");
		$im->Resize(width=>$iw, height=>$ih);
		
		my $absolute_small_filepath = File::Spec->rel2abs (
		    "$ENV{PIXENATE_CACHE_DIR}/small_$filename",
		    $ENV{PIXENATE_WORKING_DIRECTORY}
		    );
		$imres = $im->Write(filename=>$absolute_small_filepath);
	    }
	    else
	    {
		Sxoop::PXN8::Debug::log ("upload.pl> [info] max_dim of $max_dimension specified but resize not needed.\n");
		
	    }
	}
	$imres = $original_image->Write(filename=>$absolute_filepath);
	if ($imres)
	{
	    die "upload.pl> Image::Magick->Write(filename=>$absolute_filepath) failed: $imres\n";
	}
    }
    
    my $server_name = $query->server_name();
    my $server_port = $query->server_port();
    my $image_param_value = undef;
    
    my $newpage = "$next_page?";
    
    if ($resize)
    {
	$image_param_value = "$pxn8_root/$ENV{PIXENATE_CACHE_DIR}/small_$filename";
    }
    else
    {
	$image_param_value = "$pxn8_root/$relative_filepath";
    }
#
# wph 20070810 Add the servername prefix 
#
    # wph 20110531 allow for https case.
    my $protocol = "http";
    if ($ENV{HTTPS} eq "on"){
	$protocol = "https";
    }
    #
    # wph 20071205 // after the servername causes problems for some servers
    # (depends on rewrite rules on IIS)
    #
    if ($image_param_value =~ /^\//)
    {
	# begins with '/' character
	$image_param_value =  $protocol . "://$ENV{HTTP_HOST}" . $image_param_value;
    }
    else
    {
	$image_param_value = $protocol . "://$ENV{HTTP_HOST}/" . $image_param_value;
    }
    
    if ($mode eq "api")
    {
	print "Content-type: text/plain\n\n";
	print $image_param_value;
	return;
    }
    $newpage .= "$image_param_name=$image_param_value&";
    if ($hires_image_param_name)
    {
	$newpage .= "$hires_image_param_name=$pxn8_root/$relative_filepath&";
    }
    $newpage .= "originalFilename=$original_filename";
    
    foreach (keys %passthru)
    {
	$newpage .= "&$_=$passthru{$_}";
    }
    
    Sxoop::PXN8::Debug::log ("upload.pl> about to redirect to: [$newpage]\n");
    
#
# On IIS 5.1 The -nph flag must be set so that the redirect works properly.
# unfortunately, setting the nph flag breaks on Apache so I need to detect the 
# server software
#
    my $server_software = $ENV{SERVER_SOFTWARE};
#
# Don't use NPH headers on IIS6.0 
#
#	 my $use_nph = !(($server_software =~ /IIS\/6/i) or ($server_software =~ /Apache/i));
    
    #
    # wph 20090130 - turn off nph for all except IIS 5
    #
    my $use_nph = ($server_software =~ /IIS\/5/i);
    
    my $header = $query->redirect(-uri=>$newpage, -status=>302,	-nph=> $use_nph );
    
    print $header;
}

# 
# Display errors at the browser
#
sub handle_error
{
    my ($error) = @_;
    #
    # remove all newlines from error message
    #
    $error =~ s/\n//g;    $error =~ s/\r//g;
    $error =~ s/"//g;	  $error =~ s/'//g;
    
    print "Content-type: text/plain\n\n";
    print "An error occurred while trying to upload the photo.\n";
    print "$error";
    
}
